### ISSUE TITLE
False Positive: Format String Error (when the application returns a normal error message)

### ISSUE THREAD

**Issue by brittag** [zaproxy/zaproxy#4211]
My team has a web application that gets the following false positive when we run OWASP ZAP:

```
URL	https://EXAMPLE.COM/?page=ZAP%25n%25s%25n%25s%25n%25s%25n%25s%25n%25s%25n%25s%25n%25s%25n%25s%25n%25s%25n%25s%25n%25s%25n%25s%25n%25s%25n%25s%25n%25s%25n%25s%25n%25s%25n%25s%25n%25s%25n%25s%0A
Method	GET
Parameter	page
Attack	ZAP%n%s%n%s%n%s%n%s%n%s%n%s%n%s%n%s%n%s%n%s%n%s%n%s%n%s%n%s%n%s%n%s%n%s%n%s%n%s%n%s
Instances	2
Solution	
Rewrite the background program using proper deletion of bad character strings. This will require a recompile of the background executable.

Other information	
Potential Format String Error. The script closed the connection on a /%s
```

This is a false positive because the application didn't parse the string - instead, the application ignored the attempted attack and returned an error message. This is what the returned page looks like:

```
{
  "error_code": "UnknownError",
  "description": "An unknown error occurred.",
  "code": 10001
}
```

In our application, the URL `https://EXAMPLE.COM/?page=test` returns the exact same "UnknownError" error message.

To prevent this false positive, perhaps OWASP ZAP could check the results of an arbitrary parameter without an attempted Format String Error attack, and see if the application returns the same results.

Thank you!

**Comment by kingthorin** [comment]
Is the response code a 500 when the injection is sent?

<sub>Code ref: https://github.com/zaproxy/zap-extensions/blob/master/src/org/zaproxy/zap/extension/ascanrules/FormatString.java</sub>

**Comment by brittag** [comment]
Yes, the response code for that error page is a 500. The Content-Type is `text/plain`.

**Comment by psiinon** [comment]
500 means its a server error, ie the server failed to handle the request gracefully. This is exactly the sort of thing I'd want to know, and I'd argue that its a good indication that there _could_ be a problem. It all comes down to what sort of potential problems you are interested in. 
If you dont care about specific issues then you can just not use the specific rules.
If you do care about the issues in certain cases but not in others then let us know and we'll see if we can use the Threshold to allow you to choose between them.

**Comment by kingthorin** [comment]
If you send just https://EXAMPLE.COM/?page=ZAP is the response a 500?

**Comment by brittag** [comment]
@kingthorin Yes, same behavior for `https://example.com/?page=ZAP` - it returns an error page, response code 500.

I can see the reasoning that a 500 response means the application maintainer should take a close look at the application behavior to ensure that nothing unexpected happened "behind the scenes". In that case, for my use case it could help to add more explanation to the "Other information" field about the symptoms that ZAP observed.

Current:
> Other information | Potential Format String Error. The script closed the connection on a /%s

Suggested update:
> Other information | Potential Format String Error. The program closed the connection on a /%s and returned a 500 Internal Server Error. Inspect the program to determine whether the program evaluated the input string as a command.

**Comment by kingthorin** [comment]
I think I've got a way to weed out this false positive, I just discussed it on IRC and seem to have agreement.

Less specific un-handled error conditions should be caught by:

- https://github.com/zaproxy/zap-core-help/wiki/HelpAddonsAscanrulesAscanrules#parameter-tampering
- https://github.com/zaproxy/zap-core-help/wiki/HelpAddonsPscanrulesPscanrules#application-errors
- https://github.com/zaproxy/zap-extensions/wiki/HelpAddonsPscanrulesBetaPscanbeta#information-disclosure-debug-errors

**Comment by brittag** [comment]
Awesome, looking forward to trying a future release with this! 🎉  💃  ✨

**Comment by thc202** [comment]
Released in version 29 of Active scanner rules add-on.

**Comment by CatchSandeepVaid** [comment]
I am getting this error even when the return code is 401 (unauthenticated)?

**Comment by thc202** [comment]
That's for the alert message? The scanner must have seen a 500 somewhere otherwise it wouldn't raise the alert.

**Comment by ivyheasty** [comment]
I have gotten this alert the first time I scanned my application. However, on the second time running the scan, this alert was not detected. I verified the issue manually and my application returned 500 for the first 2 requests. The subsequent requests returned 200. Alert should be raised but that was not the case. I have tried setting the threshold to high, but the vulnerability still not detected.
